src/time/time.cr

```diff
@@ -567,42 +576,43 @@ struct Time
   # Time.local_offset_in_minutes # => -180
   # ```
   def self.local_offset_in_minutes
-    ifdef linux
-      -LibC.timezone.to_i32 / 60
-    else
-      if LibC.gettimeofday(nil, out tzp) != 0
-        raise Errno.new("gettimeofday")
-      end
-      -tzp.tz_minuteswest.to_i32
+    compute_ticks do |t, tz|
+      tz / Span::TicksPerMinute
     end
   end
 
   protected def self.compute_utc_ticks(ticks)
-    compute_ticks do |t, tp, tz|
-      ticks + tz
+    compute_ticks do |t, tz|
+      ticks - tz
     end
   end
 
   protected def self.compute_local_ticks(ticks)
-    compute_ticks do |t, tp, tz|
-      ticks - tz
+    compute_ticks do |t, tz|
+      ticks + tz
     end
   end
 
+  # Yields `ticks, tz`, where `ticks` is the number of ticks
+  # for "now", and `tz` is the number of ticks for now's timezone.
   private def self.compute_ticks
-    if LibC.gettimeofday(out tp, out tzp) != 0
+    LibC.tzset
+
+    # `gettimeofday` doesn't set the timezone info in the second
+    # argument from ENV["TZ"], we must use `localtime_r` which does.
+    if LibC.gettimeofday(out tp, nil) != 0
       raise Errno.new("gettimeofday")
     end
     ticks = tp.tv_sec.to_i64 * Span::TicksPerSecond + tp.tv_usec.to_i64 * 10_i64
     ticks += UnixEpoch
 
-    ifdef linux
-      tz = LibC.timezone.to_i64 / 60
-    else
-      tz = tzp.tz_minuteswest.to_i64
-    end
+    sec = tp.tv_sec
+    LibC.localtime_r(pointerof(sec), out tm)
+
+    tz = -LibC.timezone.to_i64 / 60
+    tz += 60 if tm.tm_isdst != 0
```

Comment:
BTW: what about using `offset` instead of `tz`, in order to reflect that `offset` is `tz + dst`?